socorro/unittest/schemas/test_schemas.py

```diff
@@ -25,21 +26,467 @@ def test_validate_schemas(reporoot):
+    "upload_file_minidump_browser.json_dump.threads.[].frames.[].line",
+    "upload_file_minidump_browser.json_dump.threads.[].frames.[].missing_symbols",
+    "upload_file_minidump_browser.json_dump.threads.[].frames.[].module",
+    "upload_file_minidump_browser.json_dump.threads.[].frames.[].module_offset",
+    "upload_file_minidump_browser.json_dump.threads.[].frames.[].offset",
+    "upload_file_minidump_browser.json_dump.threads.[].frames.[].registers",
+    "upload_file_minidump_browser.json_dump.threads.[].frames.[].registers.(re:^.+$)",
+    "upload_file_minidump_browser.json_dump.threads.[].frames.[].trust",
+    "upload_file_minidump_browser.json_dump.threads.[].last_error_value",
+    "upload_file_minidump_browser.json_dump.threads.[].thread_name",
+    "upload_file_minidump_browser.mdsw_return_code",
+    "upload_file_minidump_browser.mdsw_status_string",
+    "upload_file_minidump_browser.stackwalk_version",
+    "upload_file_minidump_browser.success",
+    "uptime",
+    "uptime_ts",
+    "useragent_locale",
+    "utility_process_sandboxing_kind",
+    "uuid",
+    "vendor",
+    "version",
+    "windows_error_reporting",
+    "xpcom_spin_event_loop_stack",
+}
+
+
 def test_processed_crash_schema():
     # We use the schema reducer to traverse the schema and validate the socorro metadata
     # values
 
     metadata_schema = get_file_content("socorro_metadata.1.schema.yaml")
 
+    public_fields = set()
+
     def validate_metadata(path, general_path, schema_item):
         # Print this out so it's clear which item failed
-        print(f"working on {general_path}")
         if "socorro" in schema_item:
-            # NOTE(willkg): this raises an exception if it fails and that'll trigger a
-            # pytest failure
-            jsonschema.validate(schema_item["socorro"], metadata_schema)
+            try:
+                jsonschema.validate(schema_item["socorro"], metadata_schema)
+            except Exception:
+                print(f"Exception validating {general_path}")
+                raise
+
+            permissions = schema_item["socorro"].get("permissions", [])
+
+            # Make sure all fields have permissions set
+            if not permissions:
+                raise PermissionsMissingError(f"{general_path} is missing permissions")
+
+            # Make sure all public fields have parents that are public--otherwise they
+            # will get reduced out
+            if "public" in permissions:
+                # Add this field to the public_fields set
+                public_fields.add(".".join(split_path(general_path)))
+
+                parts = list(split_path(general_path))
+                for i in range(len(parts)):
+                    parent = ".".join(parts[:i])
+                    if not parent:
+                        continue
+
+                    if parent not in public_fields:
+                        raise NonPublicParentError(
```

Comment:
Hrm... This might not work because `java_exception_raw` is protected, but it's a ref and the subschema has public fields. This should kick up an error.

Reply:
Oh, I see what's going on... yeah, this is definitely wrong. I'll fix this.